### ISSUE TITLE
Cell::setStream performance

### ISSUE THREAD

**Issue by clinuxrulz** [SodiumFRP/sodium-typescript#93]
Say we have constructed 50,000 Cells in 1 transaction.
The following lines:
```
        // We do a trick here of registering the source for the duration of the current
        // transaction so that we are guaranteed to catch any stream events that
        // occur in the same transaction.
        this.vertex.register(Vertex.NULL);
        Transaction.currentTransaction.last(() => {
            this.vertex.deregister(Vertex.NULL);
        });
```

Will add 50,000 sodium objects to the children of Vertex.NULL.
Then at the end of the transaction Vertex::decRefCount is called 50,000 times, each time removing 1 element from an child array containing initially 50,000 elements.

So time complexity O(n^2) for n Cells constructed in 1 transaction.

One work around could be to create a unique vertex for each Cell::setStream.

Like so:
```
        let tmpVertex = new Vertex("Cell::setStream", 1e12, []);
        this.vertex.register(tmpVertex);
        Transaction.currentTransaction.last(() => {
            this.vertex.deregister(tmpVertex);
        });
```

That way when Vertex::decRefCount is called 50,000 times, but the children array for call will have only 1 element.

This returns the time complexity to O(n)

**Comment by clinuxrulz** [comment]
Other option is to use a Set<> for Vertex.childrn, but may be slower than an array for a small number of elements.

**Comment by clinuxrulz** [comment]
This isn't top priority, I'm just gonna be careful with how many Cells I create.

**Comment by clinuxrulz** [comment]
The performance of "Dictionary" from "typescript-colections" is quite good. Internally it just using a javascript object. E.g.

```
let x = new Dictionary<string,string>();
x.put("a", "b");
x.put("c", "d");
```

internally just does this as a javascript object:
```
{
  $a: "b",
  $c: "d"
}
```

so the performance will be good on small number of elements too as it is optimised by the JavaScript engine (the engine swaps implementations based on collection size).

So we could go:
```
    targets: Dictionary<Vertex,number>;
    childrn: Dictionary<Vertex,number>;
```

The "number" part of like a counter because we can have the same vertex in the collection multiple times. (like a multiset from c++)

**Comment by clinuxrulz** [comment]
I believe Vertex.NULL is generally the only vertex that could have a larget number of children. I closed PR #96 without merging for that reason. There is most likely less complicated solutions to that problem.

**Comment by clinuxrulz** [comment]
Creating a unique vertex is just as good a solution as any. And appears to be the simplest. Just a shame it will hide itself when debugging memory management. Because currently we can backtrack from Vertex.NULL and reach all registered vertices.

**Comment by clinuxrulz** [comment]
And the same can be done for non-rank dependent sources.

**Comment by clinuxrulz** [comment]
fixed in rank #106
